be/queue.h

```diff
@@ -0,0 +1,198 @@
+	 * - after m0_be_get() is done the item is moved back;
+	 * - array of pointers to be_queue_item couldn't be used because the
+	 *   structure has a flexible array member at the end.
+	 */
+	char                    *bq_qitems;
+
+	/** bqop_tl, be_queue_item::bqi_link */
+	struct m0_tl             bq_op_put;
+	struct m0_tl             bq_op_put_unused;
+	struct m0_tl             bq_op_get;
+	struct m0_tl             bq_op_get_unused;
+
+	/** Is used to wait in m0_be_queue_get() */
+	struct be_queue_wait_op *bq_ops_get;
+	/** Is used to wait in m0_be_queue_put() */
+	struct be_queue_wait_op *bq_ops_put;
+
+	bool                     bq_the_end;
+
+	uint64_t                 bq_enqueued;
+	uint64_t                 bq_dequeued;
+};
+
+#define BEQ_F "(queue=%p bq_enqueued=%"PRIu64" bq_dequeued=%"PRIu64" " \
+	"bq_the_end=%d)"
+#define BEQ_P(bq) (bq), (bq)->bq_enqueued, (bq)->bq_dequeued, \
+	!!((bq)->bq_the_end)
+
+M0_INTERNAL int m0_be_queue_init(struct m0_be_queue     *bq,
+                                 struct m0_be_queue_cfg *cfg);
+M0_INTERNAL void m0_be_queue_fini(struct m0_be_queue *bq);
+
+M0_INTERNAL void m0_be_queue_lock(struct m0_be_queue *bq);
+M0_INTERNAL void m0_be_queue_unlock(struct m0_be_queue *bq);
+
+/*
+ * Put a buffer to the queue.
+ *
+ * @param data  the data from this buffer is copied to the queue internal data
+ *              structures
+ *
+ * @pre data->b_nob == bq->bq_cfg.bqc_item_length
+ *
+ * - if the queue is full, then the operation would block until a space for at
+ *   least one item becomes available in the queue;
+ * - different producers waiting on m0_be_queue_put() would be awaken in the
+ *   same order they were put to sleep.
+ */
+M0_INTERNAL void m0_be_queue_put(struct m0_be_queue  *bq,
+                                 struct m0_be_op     *op,
+                                 const struct m0_buf *data);
+
+/*
+ * Nothing is going to be put to the queue after this call.
+ * This function could be called only once.
+ */
+M0_INTERNAL void m0_be_queue_end(struct m0_be_queue *bq);
+
+/*
+ * Get a buffer from the queue.
+ *
+ * @param data          The data is copied from queue internal data structures
+ *                      to the buffer this parameter points to.
+ * @param successful    true if the data is returned, false if the queue is
+ *                      empty and m0_be_queue_end() has been called before this
+ *                      call.
+ *
+ * @pre data->b_nob == bq->bq_cfg.bqc_item_length
+ *
+ * - if the queue is empty, then the operation would block until at least one
+ *   item becomes available in the queue;
+ * - different consumers waiting on m0_be_queue_get() would be awaken in the
+ *   same order they were put to sleep.
+ */
+M0_INTERNAL void m0_be_queue_get(struct m0_be_queue *bq,
+                                 struct m0_be_op    *op,
+                                 struct m0_buf      *data,
+                                 bool               *successful);
```

Comment:
Can we make this function return a bool, so that it does not need an `bool               *successful` as its argument?

Reply:
Explained at the top of the file.